syntax = "proto2";
package RWoperation;

message KeyPoint{
    optional float angle = 1;
    optional int32 class_id = 2;
    optional int32 octave = 3;
    message point{
        optional float x = 1;
        optional float y = 2;
    }
    optional point pt = 6;
    optional float response = 4;
    optional float size = 5;
    optional string kp_name = 7;
}

message Descriptors{
    message shape{
        repeated int32 elem = 1;
    }
    optional shape dim = 1;
    repeated float element = 2 [packed=true];
    optional string des_name = 3;
}

message CombFeature{
    //save combination feature, 
    optional string name = 1;
    repeated KeyPoint kps = 2;
//    optional Descriptors des = 3;
    repeated Descriptors des = 3;
}

message Feature{
    optional string name = 1;
    repeated CombFeature feature = 2;
}

message StringDict{
    optional string key = 1;
    optional string value = 2;
}

message PathDict{
    repeated StringDict path_dict = 1;
}
